File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/java/wrapper/ChannelWrapperContext.java
package org.kevoree.library.java.wrapper;

import org.kevoree.Channel;
import org.kevoree.MBinding;
import org.kevoree.api.ChannelContext;
import org.kevoree.api.ModelService;
import org.kevoree.api.Port;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

/**
 *
 *
 */
public class ChannelWrapperContext implements ChannelContext {

    private HashMap<String, Port> boundPorts = new HashMap<>();
    private ModelService modelService;
    private String channelPath;

    public ChannelWrapperContext(String channelPath, ModelService modelService) {
        this.channelPath = channelPath;
        this.modelService = modelService;
    }

    @Override
    public List<Port> getLocalPorts() {
        return new ArrayList<>(boundPorts.values());
    }

    @Deprecated
    public HashMap<String, Port> getPortsBinded() {
        return this.getBoundPorts();
    }

    public HashMap<String, Port> getBoundPorts() {
        return boundPorts;
    }

    @Deprecated
    public void setPortsBinded(HashMap<String, Port> portsBinded) {
        this.boundPorts = portsBinded;
    }

    public String getChannelPath() {
        return channelPath;
    }

    @Override
    public List<String> getRemotePortPaths() {
        ArrayList<String> result = new ArrayList<>();
        Channel channel = (Channel) modelService.getCurrentModel().getModel().findByPath(channelPath);
        if (channel != null) {
            for (MBinding binding : channel.getBindings()) {
                org.kevoree.Port port = binding.getPort();
                if (port != null && port.getRefInParent().equals("provided")) {
                    result.add(port.path());
                }
//                ContainerNode parentNode = (ContainerNode) binding.getPort().eContainer().eContainer();
//                if (parentNode != null) {
//                    if (!parentNode.getName().equals(localNodePath)) {
//                        org.kevoree.Port finalPort = binding.getPort();
//                        if (finalPort != null) {
//                            // TODO should we just add input ports here??
//                            result.add(finalPort.path());
//                        }
//                    }
//                }
            }
        }
        return result;
    }
}

File: java/org.kevoree.library.java.javaNode/src/main/java/org/kevoree/library/java/wrapper/ChannelWrapperContext.kt
package org.kevoree.library.java.wrapper

import java.util.HashMap
import org.kevoree.api.ChannelContext
import org.kevoree.api.Port
import org.kevoree.api.ModelService
import org.kevoree.Channel
import java.util.ArrayList
import org.kevoree.ContainerNode

/**
 * Created with IntelliJ IDEA.
 * User: duke
 * Date: 28/11/2013
 * Time: 08:46
 */

public class ChannelWrapperContext(val channelPath: String, val localNodePath: String, val modelService: ModelService) : ChannelContext {

    override fun getRemotePortPaths(): MutableList<String>? {
        val result = ArrayList<String>()
        val channel = modelService.getCurrentModel()?.getModel()?.findByPath(channelPath) as? Channel
        if (channel != null) {
            for (binding in channel.bindings) {
                if ( (binding.port?.eContainer()?.eContainer() as? ContainerNode)?.name != localNodePath) {
                    val finalPort = binding.port
                    if (finalPort != null) {
                        result.add(finalPort.path())
                    }
                }
            }
        }
        return result
    }

    val portsBinded: MutableMap<String, Port> = HashMap<String, Port>()

    override fun getLocalPorts(): MutableList<Port>? {
        return portsBinded.values().toList() as MutableList<Port>
    }

}
